### PR TITLE
Feature/reactify imprecise swing

### PR DESCRIPTION
**Pull request by kzalite**

### CODE REVIEW

File: app/ts/Elicitation/ElicitationContext.tsx [hunk 1]
```diff
@@ -16,7 +17,7 @@ export function ElicitationContextProviderComponent({
 }: {
   elicitationMethod: ElicitationMethod;
   cancel: () => void;
-  save: (preferences: IExactSwingRatio[]) => void;
+  save: (preferences: (IRatioBound | IExactSwingRatio)[]) => void;
```
Comment (chinook25) [inline app/ts/Elicitation/ElicitationContext.tsx:20]: I think the typing should be 
IRatioBound[] | IExactSwingRatio[]

As it is now it allows for mixed preferences.

File: app/ts/Elicitation/ElicitationContext.tsx [hunk 2]
```diff
@@ -25,7 +26,7 @@ export function ElicitationContextProviderComponent({
     string
   >();
   const [preferences, setPreferences] = useState<
-    Record<string, IExactSwingRatio>
+    Record<string, IExactSwingRatio | IRatioBound>
```
Comment (chinook25) [inline app/ts/Elicitation/ElicitationContext.tsx:29]: Same comment: Record<string, IExactSwingRatio> | Record<string, IRatioBound>

Comment (chinook25) [reply app/ts/Elicitation/ElicitationContext.tsx:29]: Also realizing that I might not do this correctly myself all the time :P

File: app/ts/Elicitation/ImpreciseSwingElicitation/ImpreciseSwingButtons/ImpreciseSwingButtons.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
+import Button from '@material-ui/core/Button';
+import ButtonGroup from '@material-ui/core/ButtonGroup';
+import {ElicitationContext} from 'app/ts/Elicitation/ElicitationContext';
+import {PreferencesContext} from 'app/ts/Elicitation/PreferencesContext';
+import _ from 'lodash';
+import React, {useContext} from 'react';
+
+export default function ImpreciseSwingButtons() {
```
Comment (chinook25) [inline app/ts/Elicitation/ImpreciseSwingElicitation/ImpreciseSwingButtons/ImpreciseSwingButtons.tsx:8]: Isn't this file (almost) the same for precise swing? Can they be combined into one?

File: app/ts/Elicitation/ImpreciseSwingElicitation/ImpreciseSwingElicitationUtil.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+export function getImpreciseSwingStatement(
+  criterion: IElicitationCriterion
+): string {
+  return DEFAULT_PRECISE_TEMPLATE.replace(/%criterion1%/gi, criterion.title)
+    .replace(/%unit1%/gi, criterion.unitOfMeasurement.label)
+    .replace(/%worst1%/gi, String(getWorst(criterion)))
+    .replace(/%best1%/gi, String(getBest(criterion)));
+}
+
```
Comment (chinook25) [inline app/ts/Elicitation/ImpreciseSwingElicitation/ImpreciseSwingElicitationUtil.ts:15]: No test

File: app/ts/Elicitation/ImpreciseSwingElicitation/ImpreciseSwingSlider/ImpreciseSwingSlider.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    setBoundPreference(criterion.id, newValue);
+  }
+
+  return (
+    <>
+      {mostImportantCriterionId === criterion.id
+        ? '100%'
+        : `${significantDigits(sliderValue[0])} - ${significantDigits(
+            sliderValue[1]
+          )}%`}
```
Comment (chinook25) [inline app/ts/Elicitation/ImpreciseSwingElicitation/ImpreciseSwingSlider/ImpreciseSwingSlider.tsx:32]: function?

File: app/ts/Elicitation/Interface/IPreference.ts [hunk 6]
```diff
@@ -1,4 +1,9 @@
-export type ElicitationMethod = 'ranking' | 'precise' | 'matching' | 'choice';
+export type ElicitationMethod =
```
Comment (chinook25) [inline app/ts/Elicitation/Interface/IPreference.ts:1]: Lead types with a T like we do the I for interfaces. So: TElicitationMethod 
Makes it more clear what it is, but also makes collisions with types from other packages less likely.

**Review by chinook25 — CHANGES_REQUESTED** [review]

File: app/ts/Elicitation/ImpreciseSwingElicitation/ImpreciseSwingElicitationUtil.ts (hunk 4 above)
Comment (kzalite) [reply app/ts/Elicitation/ImpreciseSwingElicitation/ImpreciseSwingElicitationUtil.ts:15]: actually function is not used

**Review by chinook25 — APPROVED** [review]

File: app/ts/Elicitation/ImpreciseSwingElicitation/ImpreciseSwingElicitationUtil.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import _ from 'lodash';
+import IElicitationCriterion from '../Interface/IElicitationCriterion';
+import IRatioBound from '../Interface/IRatioBound';
+
+export function setInitialImprecisePreferences(
+  criteria: Record<string, IElicitationCriterion>,
+  mostImportantCriterionId: string
+): Record<string, IRatioBound> {
+  return _(criteria)
+    .filter((criterion) => {
```
Comment (DanielReid) [inline app/ts/Elicitation/ImpreciseSwingElicitation/ImpreciseSwingElicitationUtil.ts:10]: could be `reject(['id', mostImportantCriterionId])`

File: app/ts/Elicitation/SwingButtons/SwingButtons.tsx [hunk 8]
```diff
@@ -29,7 +32,11 @@ export default function PreciseSwingButtons() {
   }
 
   function finishElicitation() {
-    save(_.toArray(preferences));
+    if (elicitationMethod === 'imprecise') {
```
Comment (DanielReid) [inline app/ts/Elicitation/SwingButtons/SwingButtons.tsx:35]: probably forward-looking to already make this a switch over the type

**Review by DanielReid — APPROVED** [review]
Nice. Some small quibbles.

**Pull request merged**